### PR TITLE
Explicit parameter binding and other binding updates (cleaned)

### PR DESCRIPTION
**Pull request by KathleenDollard**
Unifies the binding model and provides same features (and general code path) across property, parameter and constructor parameter binding. 

This hopefully is good on the rebase issues

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,6 +1,6 @@
 
 
-[![Build Status](https://dev.azure.com/dnceng/public/_apis/build/status/dotnet/command-line-api/command-line-api?branchName=main)](https://dev.azure.com/dnceng/public/_build/latest?definitionId=337&branchName=main) [![Join the chat at https://gitter.im/dotnet/command-line-api](https://badges.gitter.im/dotnet/command-line-api.svg)](https://gitter.im/dotnet/command-line-api?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
```
Comment (jonsequitur) [inline README.md:3]: Please revert this change.

File: src/System.CommandLine.Tests/Binding/ModelBinderConstructorTests.cs [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+            var binder = new ModelBinder(typeToCreate);
+            var commandLine = $"--value {testCase.CommandLine}";
+            var bindingContext = new BindingContext(command.Parse(commandLine));
+            var instance = binder.CreateInstance(bindingContext) as TestModelBase;
+
+            instance.Value.Should().BeAssignableTo(testCase.ParameterType);
+            testCase.AssertBoundValue(instance.Value);
+        }
+
+        private static Command GetSingleArgumentCommand(BindingTestCase testCase)
```
Comment (jonsequitur) [inline src/System.CommandLine.Tests/Binding/ModelBinderConstructorTests.cs:76]: This name could be a bit more accurate, since the argument we're targeting here is actually on an option, not a command.

File: src/System.CommandLine.Tests/Binding/ModelBinderConstructorTests.cs [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+using System.IO;
+using System.Reflection;
+using System.Threading.Tasks;
+using FluentAssertions;
+using Xunit;
+using common = System.CommandLine.Tests.Binding.ModelBindingCommandHandlerTests;
+
+namespace System.CommandLine.Tests.Binding
+{
+    public class ModelBinderConstructorTests
```
Comment (jonsequitur) [inline src/System.CommandLine.Tests/Binding/ModelBinderConstructorTests.cs:16]: I see two tests in this class for option arguments but none for command arguments.

File: src/System.CommandLine.Tests/Binding/ModelBinderConstructorTests.cs [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+        [InlineData(typeof(ClassWithSetter<int>))]
+        [InlineData(typeof(ClassWithCtorParameter<string>))]
+        [InlineData(typeof(ClassWithSetter<string>))]
+        [InlineData(typeof(FileInfo))]
+        [InlineData(typeof(FileInfo[]))]
+        [InlineData(typeof(string[]))]
+        [InlineData(typeof(List<string>))]
+        [InlineData(typeof(int[]))]
+        [InlineData(typeof(List<int>))]
+        public async Task Handler_constructor_receives_option_arguments_bound_to_the_specified_type(
```
Comment (jonsequitur) [inline src/System.CommandLine.Tests/Binding/ModelBinderConstructorTests.cs:30]: `TestCommandHandler<T>` has a default constructor so the test name (and presence in the class given the class name) looks inaccurate.

File: src/System.CommandLine.Tests/Binding/ModelBindingCommandHandlerTests.cs [hunk 5]
```diff
@@ -458,6 +458,114 @@ public async Task When_binding_fails_due_to_parameter_naming_mismatch_then_handl
...
+        [InlineData(typeof(ClassWithSetter<int>))]
+        [InlineData(typeof(ClassWithCtorParameter<string>))]
+        [InlineData(typeof(ClassWithSetter<string>))]
+        [InlineData(typeof(FileInfo))]
+        [InlineData(typeof(FileInfo[]))]
+        [InlineData(typeof(string[]))]
+        [InlineData(typeof(List<string>))]
+        [InlineData(typeof(int[]))]
+        [InlineData(typeof(List<int>))]
+        public async Task Handler_method_receives_command_options_explicitly_bound_to_the_specified_type(
```
Comment (jonsequitur) [inline src/System.CommandLine.Tests/Binding/ModelBindingCommandHandlerTests.cs:527]: I think this should be named `Handler_method_receive_option_arguments_explicitly_bound_to_the_specified_type`

File: src/System.CommandLine.Tests/SuggestDirectiveTests.cs [hunk 6]
```diff
@@ -25,7 +25,7 @@ public SuggestDirectiveTests()
                 .AddSuggestions("apple", "banana", "cherry");
 
             _vegetableOption = new Option<string>("--vegetable")
-                .AddSuggestions((_, __) => new[] { "asparagus", "broccoli", "carrot" });
```
Comment (jonsequitur) [inline src/System.CommandLine.Tests/SuggestDirectiveTests.cs:28]: Please revert this.

File: src/System.CommandLine.Tests/SuggestionTests.cs [hunk 7]
```diff
@@ -170,71 +170,6 @@ public void When_an_option_has_a_default_value_it_will_still_be_suggested()
...
-
-            var result = parser.Parse("--clone  --origin test");
-
-            _output.WriteLine(result.ToString());
-
-            result.GetSuggestions(8)
-                  .Should()
-                  .BeEquivalentTo("test");
-        }
-
```
Comment (jonsequitur) [inline src/System.CommandLine.Tests/SuggestionTests.cs:237]: Please revert these test deletions.

File: src/System.CommandLine.Tests/SuggestionTests.cs [hunk 8]
```diff
@@ -622,7 +557,7 @@ public void Command_argument_suggestions_can_be_provided_using_a_delegate()
                     new Argument
                         {
                             Arity = ArgumentArity.ExactlyOne,
-                            Suggestions = { (_, __) => new[] { "vegetable", "mineral", "animal" } }
```
Comment (jonsequitur) [inline src/System.CommandLine.Tests/SuggestionTests.cs:625]: Please revert.

File: src/System.CommandLine.Tests/SuggestionTests.cs [hunk 9]
```diff
@@ -642,7 +577,7 @@ public void Option_argument_suggestions_can_be_provided_using_a_delegate()
                 {
                     Argument = new Argument<string>()
                     {
-                        Suggestions = { (_, __) => new [] { "vegetable", "mineral", "animal" } }
```
Comment (jonsequitur) [inline src/System.CommandLine.Tests/SuggestionTests.cs:645]: Please revert.

File: src/System.CommandLine/Argument.cs [hunk 10]
```diff
@@ -178,10 +178,10 @@ internal void AddAllowedValues(IEnumerable<string> values)
             AllowedValues.UnionWith(values);
         }
 
-        public override IEnumerable<string?> GetSuggestions(ParseResult? parseResult = null, string? textToMatch = null)
+        public override IEnumerable<string?> GetSuggestions(string? textToMatch = null)
         {
             var dynamicSuggestions = Suggestions
-                .SelectMany(source => source.GetSuggestions(parseResult, textToMatch));
```
Comment (jonsequitur) [inline src/System.CommandLine/Argument.cs:184]: Please revert the changes to this file.

File: src/System.CommandLine/Binding/BoundValue.cs [hunk 11]
```diff
@@ -53,4 +53,26 @@ public static BoundValue DefaultForValueDescriptor(IValueDescriptor valueDescrip
...
+    //        {
+    //            if (!_valueHasBeenSet )
+    //            {
+    //                object? value;
+    //               if (ValueSource.TryGetValue(ValueDescriptor, ))
+    //            }
+    //            return base.Value;
+    //        }
+    //    }
+    //}
```
Comment (jonsequitur) [inline src/System.CommandLine/Binding/BoundValue.cs:77]: You can delete the commented out code.

File: src/System.CommandLine/Binding/BoundValue.cs [hunk 12]
```diff
@@ -25,7 +25,7 @@ public class BoundValue
 
         public IValueSource ValueSource { get; }
 
-        public object? Value { get; }
+        public virtual object? Value { get; }
```
Comment (jonsequitur) [inline src/System.CommandLine/Binding/BoundValue.cs:28]: Does this need to be `virtual`?

File: src/System.CommandLine/Binding/ModelBinder.cs [hunk 13]
```diff
@@ -1,34 +1,23 @@
-// Copyright (c) .NET Foundation and contributors. All rights reserved.
-// Licensed under the MIT license. See LICENSE file in the project root for full license information.
```
Comment (jonsequitur) [inline src/System.CommandLine/Binding/ModelBinder.cs:2]: Please put back the copyright header.

File: src/System.CommandLine/Binding/ModelBinder.cs [hunk 14]
```diff
@@ -37,208 +26,210 @@ internal ModelBinder(IValueDescriptor valueDescriptor)
...
+            var explicitTypesToShortcut = new List<Type>
+            {
+                typeof(string)
+            };
+            Type modelType = ModelDescriptor.ModelType;
+            return modelType.IsPrimitive ||
+                   IsNullable(modelType) ||
+                   explicitTypesToShortcut.Contains(modelType);
+
+            static bool IsNullable(Type type)
```
Comment (jonsequitur) [inline src/System.CommandLine/Binding/ModelBinder.cs:108]: We have this method in `Binder.cs` unless it was deleted from there.

File: src/System.CommandLine/Binding/ModelBinder.cs [hunk 15]
```diff
@@ -37,208 +26,210 @@ internal ModelBinder(IValueDescriptor valueDescriptor)
...
+                newInstance = constructor.Invoke(values);
             }
-
-            instance = null;
-            return false;
+            catch
+            {
+                return (false, null, false);
+            }
+            if (!(newInstance is null))
```
Comment (jonsequitur) [inline src/System.CommandLine/Binding/ModelBinder.cs:139]: I believe `newInstance` can never be null.

File: src/System.CommandLine/Invocation/CommandHandler.cs [hunk 16]
```diff
@@ -46,22 +46,6 @@ public static class CommandHandler
...
-            HandlerDescriptor.FromDelegate(action).GetCommandHandler();
-
-        public static ICommandHandler Create<T1, T2, T3, T4, T5, T6, T7, T8, T9, T10>(
-            Action<T1, T2, T3, T4, T5, T6, T7, T8, T9, T10> action) =>
-            HandlerDescriptor.FromDelegate(action).GetCommandHandler();
-
-        public static ICommandHandler Create<T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11>(
-            Action<T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11> action) =>
-            HandlerDescriptor.FromDelegate(action).GetCommandHandler();
-
```
Comment (jonsequitur) [inline src/System.CommandLine/Invocation/CommandHandler.cs:64]: Please revert.

File: src/System.CommandLine/Invocation/InvocationExtensions.cs [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+        public static void BindParameter(this ICommandHandler handler, ParameterInfo param, Argument argument)
+        {
+            if (!(handler is ModelBindingCommandHandler bindingHandler))
+            {
+                throw new InvalidOperationException("Cannot bind to this type of handler");
+            }
+            bindingHandler.BindParameter(param, argument);
+        }
+    }
+}
```
Comment (jonsequitur) [inline src/System.CommandLine/Invocation/InvocationExtensions.cs:26]: I believe we discussed removing these extension methods.

File: src/System.CommandLine/Invocation/ModelBindingCommandHandler.cs [hunk 18]
```diff
@@ -15,59 +15,59 @@ internal class ModelBindingCommandHandler : ICommandHandler
         private readonly object? _invocationTarget;
         private readonly ModelBinder? _invocationTargetBinder;
         private readonly MethodInfo? _handlerMethodInfo;
-        private readonly IReadOnlyList<ParameterDescriptor> _parameterDescriptors;
+        private readonly IMethodDescriptor _methodDescriptor;
+        private Dictionary<IValueDescriptor, IValueSource> invokeArgumentBindingSources { get; } =
+            new Dictionary<IValueDescriptor, IValueSource>();
+        private readonly bool EnforceExplicitBinding = false; // ?? Wrong formatting as hint to figure out how to set this
```
Comment (jonsequitur) [inline src/System.CommandLine/Invocation/ModelBindingCommandHandler.cs:21]: This should be named `_enforceExplicitBinding`.

File: src/System.CommandLine/Parsing/ParseResultExtensions.cs [hunk 19]
```diff
@@ -216,7 +216,7 @@ public static string Diagram(this ParseResult result)
 
             var currentSymbolSuggestions =
                 currentSymbol is ISuggestionSource currentSuggestionSource
-                    ? currentSuggestionSource.GetSuggestions(parseResult, textToMatch)
```
Comment (jonsequitur) [inline src/System.CommandLine/Parsing/ParseResultExtensions.cs:219]: Please revert.

File: src/System.CommandLine/Parsing/ParseResultExtensions.cs [hunk 20]
```diff
@@ -230,7 +230,7 @@ public static string Diagram(this ParseResult result)
             else
             {
                 siblingSuggestions = parentSymbol
-                                     .GetSuggestions(parseResult, textToMatch)
```
Comment (jonsequitur) [inline src/System.CommandLine/Parsing/ParseResultExtensions.cs:233]: Please revert.

File: src/System.CommandLine/SuggestionSourceExtensions.cs [hunk 21]
```diff
@@ -39,7 +39,7 @@ public static class SuggestionSourceExtensions
                 throw new ArgumentNullException(nameof(suggestions));
             }
 
-            suggestionSources.Add(new AnonymousSuggestionSource((_, __) => suggestions));
```
Comment (jonsequitur) [inline src/System.CommandLine/SuggestionSourceExtensions.cs:42]: Please revert.

File: src/System.CommandLine/Suggestions/AnonymousSuggestionSource.cs [hunk 22]
```diff
@@ -15,9 +14,9 @@ public AnonymousSuggestionSource(SuggestDelegate suggest)
             _suggest = suggest ?? throw new ArgumentNullException(nameof(suggest));
         }
 
-        public IEnumerable<string> GetSuggestions(ParseResult? parseResult = null, string? textToMatch = null)
```
Comment (jonsequitur) [inline src/System.CommandLine/Suggestions/AnonymousSuggestionSource.cs:18]: Please revert.

File: src/System.CommandLine/Suggestions/AnonymousSuggestionSource.cs [hunk 23]
```diff
@@ -15,9 +14,9 @@ public AnonymousSuggestionSource(SuggestDelegate suggest)
...
+        public IEnumerable<string> GetSuggestions(string? textToMatch = null)
         {
-            return _suggest(parseResult, textToMatch);
```
Comment (jonsequitur) [inline src/System.CommandLine/Suggestions/AnonymousSuggestionSource.cs:20]: Please revert.

File: src/System.CommandLine/Suggestions/ISuggestionSource.cs [hunk 24]
```diff
@@ -2,12 +2,11 @@
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
 using System.Collections.Generic;
-using System.CommandLine.Parsing;
 
 namespace System.CommandLine.Suggestions
 {
     public interface ISuggestionSource
     {
-        IEnumerable<string?> GetSuggestions(ParseResult? parseResult = null, string? textToMatch = null);
```
Comment (jonsequitur) [inline src/System.CommandLine/Suggestions/ISuggestionSource.cs:11]: Please revert.

File: src/System.CommandLine/Suggestions/SuggestDelegate.cs [hunk 25]
```diff
@@ -2,9 +2,8 @@
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
 using System.Collections.Generic;
-using System.CommandLine.Parsing;
 
 namespace System.CommandLine.Suggestions
 {
-    public delegate IEnumerable<string> SuggestDelegate(ParseResult? parseResult, string? textToMatch);
```
Comment (jonsequitur) [inline src/System.CommandLine/Suggestions/SuggestDelegate.cs:9]: Please revert.

File: src/System.CommandLine/Suggestions/SuggestionSource.cs [hunk 26]
```diff
@@ -31,18 +31,18 @@ ISuggestionSource CreateForType(Type t)
...
                 {
-                    return new AnonymousSuggestionSource((_, __) => _trueAndFalse);
+                    return new AnonymousSuggestionSource(_ => _trueAndFalse);
                 }
 
                 return Empty;
             }
         }
 
-        public static ISuggestionSource Empty { get; } = new AnonymousSuggestionSource((_, __) => Array.Empty<string>());
```
Comment (jonsequitur) [inline src/System.CommandLine/Suggestions/SuggestionSource.cs:46]: You can revert this whole file.

File: src/System.CommandLine/Symbol.cs [hunk 27]
```diff
@@ -137,12 +137,12 @@ public bool HasAlias(string alias)
 
         public bool IsHidden { get; set; }
 
-        public virtual IEnumerable<string?> GetSuggestions(ParseResult? parseResult = null, string? textToMatch = null)
+        public virtual IEnumerable<string?> GetSuggestions(string? textToMatch = null)
         {
             var argumentSuggestions =
                 Children
                     .OfType<IArgument>()
-                    .SelectMany(a => a.GetSuggestions(parseResult, textToMatch))
```
Comment (jonsequitur) [inline src/System.CommandLine/Symbol.cs:145]: You can revert this whole file.

File: src/System.CommandLine/SymbolExtensions.cs [hunk 28]
```diff
@@ -33,10 +33,5 @@ internal static IEnumerable<IArgument> Arguments(this ISymbol symbol)
                     throw new NotSupportedException();
             }
         }
-
-        public static IEnumerable<string?> GetSuggestions(this ISymbol symbol, string? textToMatch = null)
-        {
-            return symbol.GetSuggestions(null, textToMatch);
-        }
     }
```
Comment (jonsequitur) [inline src/System.CommandLine/SymbolExtensions.cs:36]: You can revert this whole file.

File: src/System.CommandLine.Tests/Binding/ModelBinderConstructorTests.cs (hunk 4 above)
Comment (KathleenDollard) [reply src/System.CommandLine.Tests/Binding/ModelBinderConstructorTests.cs:30]: I deleted this file because it was identical to tests in ModelBindingCommandHandlerTests. 

We could use more constructor tests, but I think this should correspond to updating the method of binding tests.

**Review by jonsequitur — APPROVED** [review]
